src/udp.rs

```diff
@@ -1,36 +1,111 @@
-use crate::bus::Bus;
-use crate::device::{Device, DeviceRefMut};
-use crate::host::Host;
-use crate::register::socketn;
-use crate::socket::Socket;
-use core::fmt::Debug;
+use core::{convert::TryFrom, fmt::Debug};
+
 use embedded_nal::{nb, IpAddr, Ipv4Addr, SocketAddr, SocketAddrV4, UdpClientStack, UdpFullStack};
 
+use crate::{
+    bus::Bus,
+    device::{Device, DeviceRefMut},
+    host::Host,
+    register::socketn::{self, Status},
+    socket::Socket,
+};
+
+/// W5500 UDP Header
+///
+/// Contains the destination IP address, port and the length of the data.
+///
+/// Packet frame, as described in W5200 docs section 5.2.2.1:
+///
+/// ```text
+/// |<-- read_pointer                                 read_pointer + received_size -->|
+/// | Destination IP Address | Destination Port | Byte Size of DATA | Actual DATA ... |
+/// |    --- 4 Bytes ---     |  --- 2 Bytes --- |  --- 2 Bytes ---  |      ....       |
+/// ```
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct UdpHeader {
+    /// The origin socket address (IP address and port).
+    #[cfg_attr(feature = "defmt", defmt(Debug2Format))]
+    pub origin: SocketAddrV4,
+    /// Length of the UDP packet in bytes.
+    ///
+    /// This may not be equal to the length of the data in the socket buffer if the UDP packet was truncated
+    /// due to small buffer or the size of the internal W5500 RX buffer for the socket.
+    pub len: usize,
+}
+
+impl UdpHeader {
+    pub fn from_array(array: [u8; 8]) -> Self {
+        Self::from(array)
+    }
+}
+
+impl From<[u8; 8]> for UdpHeader {
+    fn from(header: [u8; 8]) -> Self {
+        let origin_addr = SocketAddrV4::new(
+            Ipv4Addr::new(header[0], header[1], header[2], header[3]),
+            u16::from_be_bytes([header[4], header[5]]),
+        );
+
+        let packet_len: usize = u16::from_be_bytes([header[6], header[7]]).into();
+
+        Self {
+            origin: origin_addr,
+            len: packet_len,
+        }
+    }
+}
+
 #[derive(Debug)]
 #[cfg_attr(feature = "defmt", derive(defmt::Format))]
 pub struct UdpSocket {
     socket: Socket,
+    /// Whether or not there has been a destination set for the socket.
+    #[cfg_attr(feature = "defmt", defmt(Debug2Format))]
+    destination: Option<SocketAddrV4>,
+    /// The local port of the socket
+    port: u16,
 }
 
 impl UdpSocket {
     fn new(socket: Socket) -> Self {
-        UdpSocket { socket }
+        let socket_index = socket.index;
+        UdpSocket {
+            socket,
+            destination: None,
+            // TODO dynamically select a random port
+            // chosen by fair dice roll.
+            // guaranteed to be random.
```

Comment:
Can you spawn an issue to handle this? I had to do something for tcp in the `smoltcp-nal`